crates/agglayer-storage/src/storage/backup/mod.rs

```diff
@@ -0,0 +1,327 @@
+                    }
+                }
+            }
+        } else {
+            if let Err(error) = self
+                .state_engine
+                .create_new_backup_flush(&self.state_db.rocksdb, true)
+            {
+                error!("Failed to create backup for state db: {:?}", error);
+            }
+
+            if let Err(error) = self
+                .state_engine
+                .purge_old_backups(self.state_max_backup_number)
+            {
+                error!("Failed to purge old backup for state db: {:?}", error);
+            }
+
+            if let Err(error) = self
+                .pending_engine
+                .create_new_backup_flush(&self.pending_db.rocksdb, true)
+            {
+                error!("Failed to create backup for pending db: {:?}", error);
+            }
+
+            if let Err(error) = self
+                .pending_engine
+                .purge_old_backups(self.pending_max_backup_number)
+            {
+                error!("Failed to purge old backup for pending db: {:?}", error);
+            }
+        }
+
+        info!("Backup successfully created");
+
+        Ok(())
+    }
+
+    /// Run the backup engine, listen for new backup requests.
+    pub async fn run(mut self, cancellation_token: CancellationToken) -> Result<(), BackupError> {
+        loop {
+            tokio::select! {
+                _ = cancellation_token.cancelled() => {
+                    info!("Backup engine cancelled");
+                    break;
+                }
+                Some(request) = self.backup_request.recv() =>{
+                    self.create_new_backup(&request)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Restore the state database from the latest backup.
+    pub fn restore(path: &Path, db_path: &Path) -> Result<(), BackupError> {
+        let env = rocksdb::Env::new()?;
+        let opts = rocksdb::backup::BackupEngineOptions::new(path)?;
+
+        let mut engine = RocksBackupEngine::open(&opts, &env)?;
+
+        std::fs::create_dir_all(db_path).unwrap();
+
+        Ok(engine.restore_from_latest_backup(db_path, db_path, &RestoreOptions::default())?)
+    }
+
+    /// Restore the state database from the defined backup version.
+    pub fn restore_at(path: &Path, db_path: &Path, version: u32) -> Result<(), BackupError> {
+        let env = rocksdb::Env::new()?;
+        let opts = rocksdb::backup::BackupEngineOptions::new(path)?;
+
+        let mut engine = RocksBackupEngine::open(&opts, &env)?;
+
+        std::fs::create_dir_all(db_path).unwrap();
+
+        Ok(engine.restore_from_backup(db_path, db_path, &RestoreOptions::default(), version)?)
+    }
+
+    pub fn list_backups(path: &Path) -> Result<BackupReport, BackupError> {
+        let env = rocksdb::Env::new()?;
+
+        let mut report = BackupReport::default();
+
+        let opts = BackupEngineOptions::new(path.join("state"))?;
```

Comment:
Minor: Would probably be better to have some sort of `BackupConfig` object instead of bare path. The object can have methods like `state_backup_path`, `epoch_backup_path`, etc. That would avoid the repeated use of string literals which are somewhat error prone and harder to maintain.